### PR TITLE
[demo] Server: Add accuracy tracking

### PR DESCRIPTION
**Pull request by hkaur008**
#124

### CODE REVIEW

**Comment by hkaur008** [comment]
#92  table to track accuracy is created now I need to insert the timestamp , transaction hash , accuracy and model id .
i can create a separate function to insert values to the table and run that function when a particular API is called .

File: demo/server.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ initSqlJs().then(SQL => {
     sqlstr = "CREATE TABLE model (id INTEGER PRIMARY KEY, name TEXT, address TEXT, description TEXT, model_type TEXT, encoder TEXT, accuracy NUMBER);"
       + "CREATE TABLE data (transaction_hash TEXT PRIMARY KEY, text TEXT);"
       + "CREATE INDEX index_address ON model(address);";
```
Comment (juharris) [inline demo/server.js:22]: You'll need to remove the ";" at the end of this line.

**Comment by juharris** [comment]
Would you add two new functions: 1) get the accuracy history for a model, 2) Add a new accuracy record for a model

File: demo/server.js [hunk 2]
```diff
@@ -20,6 +20,7 @@ initSqlJs().then(SQL => {
...
+      + "CREATE Table accuracy_time ( transaction_hash TEXT ,id INTEGER ,accuracy NUMBER , Timestamp DATETIME DEFAULT CURRENT_TIMESTAMP); "
```
Comment (juharris) [inline demo/server.js:23]: Thanks for setting this up!
Please keep the style consistent with the other statements. So it should be "CREATE TABLE ..." and use all lowercase column names.
As for some of these names. For the table name, just "accuracy" is fine.
I'm not sure what the purpose of the transaction_hash column is so it would be good to add a comment in the code to explain it to people when they read the code. The transaction_hash is the data table is the transcation hash in the Ethereum blockchain for when the data was added. I'm not sure if the same thing is appropriate here since the accuracy can be computed off-chain. Instead (or in addition), having a column for the block number would be good so we know as of which block the model has the recorded accuracy. 
I guess "id" is the model ID? Would you call it "model_id" and use a FOREIGN KEY: `FOREIGN KEY (model_id) REFERENCES model (id)`. See https://www.sqlitetutorial.net/sqlite-foreign-key/ for details.
"Timestamp" should be lowercase.

Sorry for sounding so picky. I'm always like this. I'm just trying to teach you some things.

Comment (hkaur008) [reply demo/server.js:23]: no worries, I understand. I do all the changes!

File: demo/server.js (hunk 1 above)
Comment (hkaur008) [reply demo/server.js]: done !

File: demo/server.js [hunk 3]
```diff
@@ -19,7 +19,8 @@ initSqlJs().then(SQL => {
     db = new SQL.Database();
     sqlstr = "CREATE TABLE model (id INTEGER PRIMARY KEY, name TEXT, address TEXT, description TEXT, model_type TEXT, encoder TEXT, accuracy NUMBER);"
       + "CREATE TABLE data (transaction_hash TEXT PRIMARY KEY, text TEXT);"
-      + "CREATE INDEX index_address ON model(address);";
+      + "CREATE INDEX index_address ON model(address);"
+      + "CREATE TABLE time_accuracy ( transaction_hash TEXT,block_no INTEGER  , model_id INTEGER ,accuracy NUMBER , timestamp DATETIME DEFAULT CURRENT_TIMESTAMP ,FOREIGN KEY (model_id) REFERENCES model (id)); ";
```
Comment (juharris) [inline demo/server.js:23]: Let's use clearer names. So how about "block_number" instead of "block_no"?
How about just "accuracy" for the table name?
Also would you make the styling consistent with the rest of the command so that it's easier to read: no space after "(", no spaces before ",", just one space after ",".

Comment (hkaur008) [reply demo/server.js:23]: okay sure

**Comment by hkaur008** [comment]
I have created a function to 2.) add new accuracy 
ca you direct me how exactly I need to build get accuracy record function !

File: demo/server.js [hunk 4]
```diff
@@ -151,3 +152,24 @@ initSqlJs().then(SQL => {
...
+ {
+
+ }
+ 
+ // Add a new accuracy record for a model
+  function addAccuracyRecord(accuracy)
+  {
+   db.run('INSERT INTO accuracy VALUES (?, ?, ?, ?);', [
+     accuracy.transaction_hash,
+     accuracy.block_number,
```
Comment (juharris) [inline demo/server.js:167]: These keys should be camelCase because the come JavaScript or JSON in an API.

Comment (hkaur008) [reply demo/server.js]: Done !

**Comment by juharris** [comment]
Yeah that's right. It would be good to be able to get the entire accuracy history of a model by providing a `modelId` OR an `address`.

**Comment by hkaur008** [comment]
I had built the function and ordered it by timestamp and return the result as an object. Do let me know the further changes required. I will be happy to code

File: demo/server.js [hunk 5]
```diff
@@ -151,3 +152,28 @@ initSqlJs().then(SQL => {
...
+     accuracy.modelId,
+     accuracy.accuracy,
+   ]);
+   fs.writeFile(dbPath, Buffer.from(db.export()), () => { });
+  }
+
+ // Get the accuracy history for a model
+ function getAccuracyHistory(modelId)
+ {
+  const getAccuracyStmt = db.prepare('SELECT * FROM accuracy WHERE model_id == $modelId ORDER BY timestamp;');
```
Comment (juharris) [inline demo/server.js:171]: It would also be good to support searching by model address.
The method should return object with camelCase style keys.

Comment (hkaur008) [reply demo/server.js:171]: I can make one more function to search by model address.
for this thing we need to figure out model_id by model address in the model table 
and then the select rows with that model_id in the accuracy table .

We can do this in one SQLite command as well to support searching by model address.

Comment (juharris) [reply demo/server.js:171]: You can search by model address in this function too and have the `SELECT` do a `JOIN` and use `OR` in the search.

File: demo/server.js [hunk 6]
```diff
@@ -151,3 +152,28 @@ initSqlJs().then(SQL => {
...
+  const result = getAccuracyStmt.getAsObject({ $modelId: modelId });
```
[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: image
client side results after running

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: image
heath api also running !

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: image
all models i found ! with get all

[the remark this answers is not in the document]
Comment (juharris) [reply demo/server.js:172]: So the existing methods work, that's not surprising since they haven't changed. What about your new method, it would be good to see that you can actual get the history of accuracies from your new table because the documentation for `getAsObject` makes it seem like the current implementation will not work.

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: > So the existing methods work, that's not surprising since they haven't changed. What about your new method, it would be good to see that you can actual get the history of accuracies from your new table because the documentation for `getAsObject` makes it seem like the current implementation will not work.

I was going to talk about it, I created a function to add data in the accuracy table but had not called it anywhere so no new data is being added in the accuracy table. We are having 3 models. where should I call this function (addAccuracyHistory) so that its data gets added to the accuracy table

[the remark this answers is not in the document]
Comment (juharris) [reply demo/server.js:172]: You can start by making an endpoint, `app.post('/api/accuracy', jsonParser, (req, res) => { ...` and calling you new method there.
Also make this following to retrieve the accuracy history:
```JavaScript
app.get('/api/accuracy', (req, res) => {
    const { modelId, address } = req.query;
...
```

As for where to use those, we can do that the client/ folder but we'll do that in another pull request.

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: image
This is the get API  i created ,
if db is already created then i have to run the  add table accuracy command in if to add new table , because it will load the previous db which don't have accuracy table. I have added command in else block to create new db

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js:172]: image
`getAsObject` is giving only one object of the table , I had made the result on console

[the remark this answers is not in the document]
Comment (juharris) [reply demo/server.js]: Right, so the `getAccuracyHistory` method is not very useful as it is because its name is misleading and it only gets one row.

[the remark this answers is not in the document]
Comment (hkaur008) [reply demo/server.js]: Yeah, we can do all this in one get Api

File: demo/server.js [hunk 7]
```diff
@@ -1,14 +1,14 @@
-const express = require('express');
-const fs = require('fs');
-const initSqlJs = require('sql.js');
+const express = require("express");
```
Comment (juharris) [inline demo/server.js:1]: Use double quotes for text that people will read like messages and log statments. Use single quotes for text that is for machine processing.

File: demo/server.js [hunk 8]
```diff
@@ -43,62 +50,70 @@ initSqlJs().then(SQL => {
   }
 
   // Health
-  app.get('/api/health', (req, res) => {
+  app.get("/api/health", (req, res) => {
     res.send({ healthy: true });
-  })
+  });
```
Comment (juharris) [inline demo/server.js:55]: Please don't make changes that aren't relevant to this pull request.

File: demo/server.js [hunk 9]
```diff
@@ -126,28 +141,79 @@ initSqlJs().then(SQL => {
...
+  }
+ // Add a new accuracy record for a model
+  app.post("/api/accuracy", jsonParser, (req, res) => {
+    const body = req.body;
+    presistAccuracyRecord(body);
+    return res.sendStatus(200);
+  });
+
+  // Get the accuracy history for a model
+  function getAccuracyHistory(modelId) {
```
Comment (juharris) [inline demo/server.js:191]: Delete this method because it is not used and I don't think it will work so I believe it is misleading.

File: demo/server.js [hunk 10]
```diff
@@ -126,28 +141,79 @@ initSqlJs().then(SQL => {
...
+    );
+    const result = getAccuracyStmt.getAsObject({ $modelId: modelId });
+    getAccuracyStmt.free();
+    console.log(result);
+    return result;
+  }
+  
+  // Get the accuracy history 
+  app.get("/api/accuracy", (req, res) => {
+    const getStmt = db.prepare("SELECT * FROM accuracy");
```
Comment (juharris) [inline demo/server.js:203]: We should only get the accuracy for a specific modelId.

File: demo/server.js (hunk 8 above)
Comment (hkaur008) [reply demo/server.js:55]: Sorry , actually the text got formatted automatically by the Visual code extension, had not made any chages. Let me revert changes .

File: demo/server.js [hunk 11]
```diff
@@ -1,10 +1,10 @@
 const express = require('express');
 const fs = require('fs');
+const { waitForDebugger } = require('inspector');
```
Comment (juharris) [inline demo/server.js:3]: Delete this line. It's fine if you do this locally but you should not commit debugger stuff.

File: demo/server.js [hunk 12]
```diff
@@ -150,4 +150,46 @@ initSqlJs().then(SQL => {
...
+  });
+
+  // Get the accuracy history 
+  app.get('/api/accuracy/model', (req, res) => {
+    const { modelId  } = req.query;
+    if (modelId != null) {
+      const str = 'SELECT * FROM accuracy where model_id == ' +modelId+' ORDER BY timestamp;';
+      const getStmt = db.prepare(str);
+      const accuracyHistory = [];
+    while (getStmt.step()) {
```
Comment (juharris) [inline demo/server.js:179]: The indentation is wrong on this line.

File: demo/server.js [hunk 13]
```diff
@@ -150,4 +150,46 @@ initSqlJs().then(SQL => {
...
+    const body = req.body
+    presistAccuracyRecord(body);
+    return res.sendStatus(200);
+  });
+
+  // Get the accuracy history 
+  app.get('/api/accuracy/model', (req, res) => {
+    const { modelId  } = req.query;
+    if (modelId != null) {
+      const str = 'SELECT * FROM accuracy where model_id == ' +modelId+' ORDER BY timestamp;';
```
Comment (juharris) [inline demo/server.js:176]: VERY INSECURE.

File: demo/server.js [hunk 14]
```diff
@@ -150,4 +150,46 @@ initSqlJs().then(SQL => {
...
+        blockNumber: accuracy[1],
+        modelId: accuracy[2],
+        accuracy: accuracy[3],
+        timestamp: accuracy[4],
+      });
+    }
+    getStmt.free();
+    res.send({ accuracyHistory});
+    } else {
+      return res.status(400).send({ message: "Not found." });
```
Comment (juharris) [inline demo/server.js:192]: I guess you copied this line from elsewhere in the file but the message is not really true. The problem was that the modelId was not given in the request.

Comment (juharris) [reply demo/server.js:192]: Well it could have been given as `null` but the modelId was not given as a valid value.

File: demo/server.js (hunk 13 above)
Comment (hkaur008) [reply demo/server.js:176]: then have to find a different method !

File: demo/server.js (hunk 14 above)
Comment (hkaur008) [reply demo/server.js:192]: > I guess you copied this line from elsewhere in the file but the message is not really true. The problem was that the modelId was not given in the request.

yeah this copied from your above code to find model by modelled or address , I change it to null

**Comment by hkaur008** [comment]
Results so obtained 
image
image
image

**Comment by juharris** [comment]
Please let me resolve the issues.

File: demo/server.js [hunk 15]
```diff
@@ -150,4 +150,52 @@ initSqlJs().then(SQL => {
...
+        });
+      }
+      getStmt.free();
+      if (accuracyHistory.length) {
+        res.send({ accuracyHistory });
+      } else {
+        res.send({ message: "Not Valid modelId" });
+      }
+    } else {
+      return res.status(400).send({ message: "Null" });
```
Comment (juharris) [inline demo/server.js:198]: The message should be clearer about what the issue was.

Comment (hkaur008) [reply demo/server.js:198]: okay i try to improve the messages

**Comment by hkaur008** [comment]
Messages updated !

**Comment by hkaur008** [comment]
> Messages updated !

Please review my pull request  @juharris Sir !

File: demo/server.js [hunk 16]
```diff
@@ -150,4 +150,52 @@ initSqlJs().then(SQL => {
...
+    db.run('INSERT INTO accuracy VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP);', [
+      accuracy.transactionHash,
+      accuracy.blockNumber,
+      accuracy.modelId,
+      accuracy.accuracy,
+    ]);
+    fs.writeFile(dbPath, Buffer.from(db.export()), () => {});
+  }
+  
+ // Add a new accuracy record for a model
```
Comment (juharris) [inline demo/server.js:165]: There's a few styling inconsistencies that I didn't want to buy you about changing. Your editor would handle most of these with automatic formatting. Instead of bugging you to make those changes, I set up some tools for automatic linting. #127 adds support for `yarn lint-fix` so once you pull in main, then you should run `yarn lint-fix` and it should resolve most issues automatically or tell you where there is errors. Then I can finally merge this!

Comment (juharris) [reply demo/server.js:165]: Oh I commented on this line because there's an issue with it: 1 space at the beginning instead of 2. Yeah it's a small thing but keeping the style clean and consistent is important for easily maintaining the code. Anyway, the automatic linting should resolve this issue.

Comment (hkaur008) [reply demo/server.js:165]: Okay , i change it now !

Comment (hkaur008) [reply demo/server.js]: Hey , used lint fix it fixed all the things itself , you can check now

File: demo/server.js [hunk 17]
```diff
@@ -130,7 +130,6 @@ initSqlJs().then(SQL => {
 			data.transactionHash,
 			data.originalData.text,
 		])
-		// eslint-disable-next-line @typescript-eslint/no-empty-function
```
Comment (juharris) [inline demo/server.js:133]: Why did you delete this line?

Comment (hkaur008) [reply demo/server.js:133]: automation lint fix did

Comment (hkaur008) [reply demo/server.js:133]: i thought you have added this feature in  fix also to delete eslint

Comment (juharris) [reply demo/server.js:133]: I'm confused. Are you saying that `yarn lint-fix` removed that line? It did not for me.

> I thought you have added this feature in fix also to delete eslint
What does that mean?

Please add the line back.

Comment (hkaur008) [reply demo/server.js:133]: okay
sure
